### ISSUE TITLE
sphinx can nit copied specific part from an old project 

### ISSUE THREAD

**Issue by mostafajoulaian** [pyiron/pyiron#1262]
I use the latest version of pyiron which is installed on the cluster.
I try to recover or coping some important setting from an old project (such as energy cut off, K point sampling, etc.). and try to run Murnaghan calculation. (for sphinx)
 It works very well, but what is not work is setting up a sphinx reference job with all the settings I want to make. it is deeper than the sphinx interior. Such as checking overlap.
All setting is set not in generic part, but sphinx specific part is not copied. In conclusion, no matter what you are doing in a reference job, it will never appear in the job is ultimately run.
Here is the piece of code 

def launch_sphinx_vasp (old_project):
    old_ref=old_project.load("strain_1_0")
    elem=old_ref.initial_structure.elements[0].symbol
    if isinstance(old_ref.input.KpointFolding,Group):
        kpoints=  [k for k in old_ref.input.KpointFolding.values()]
    else:
        kpoints=old_ref.input.KpointFolding 
    job_name=elem +  '_k' + 'x'.join([str(k) for k in kpoints])
    sphinx_vasp_ref = new_project.create_job(new_project.job_type.Sphinx, job_name +'_sphinx_vasp')
    sphinx_vasp_ref.structure = old_ref.initial_structure
    sphinx_abinit_ref.calc_static()
    #print(sphinx_vasp_ref.input.sphinx.to_sphinx())
    sphinx_vasp_ref.input['KpointCoords'] = [0.0, 0.0, 0.0]
    sphinx_vasp_ref.set_kpoints(kpoints)
    sphinx_vasp_ref.set_encut(old_ref.encut)
    sphinx_vasp_ref.input['EmptyStates'] = old_ref.input['EmptyStates']
    sphinx_vasp_ref.input['KJxc'] = True 
    sphinx_vasp_ref.server.queue = 'cm'
    sphinx_vasp_ref.server.cores = old_ref.server.cores
    sphinx_vasp_ref.set_occupancy_smearing(width=0.1)
    sphinx_vasp_ref.input.sphinx.pawPot.species={"checkOverlap":0}
    
    
    ########
    murn_sphinx_vasp = new_project.create_job(new_project.job_type.Murnaghan, job_name+str("_sphinx_vasp_murn"))
    murn_sphinx_vasp.input['vol_range'] = 0.06
    murn_sphinx_vasp.input['num_points'] = 7
    murn_sphinx_vasp.ref_job = sphinx_vasp_ref
    murn_sphinx_vasp.run()

One speculation is that the sphinx job is copied, loses the sphinx specific information, or at the way; jobs are set up in Murnaghan job are not compatible with coping sphinx job.
@ashtonmv @freyso

**Comment by samwaseda** [comment]
> I use the latest version of pyiron which is installed on the cluster.
> I try to recover or coping some important setting from an old project (such as energy cut off, K point sampling, etc.). and try to run Murnaghan calculation. (for sphinx)
> It works very well, but what is not work is setting up a sphinx reference job with all the settings I want to make. it is deeper than the sphinx interior. Such as checking overlap.
> All setting is set not in generic part, but sphinx specific part is not copied. In conclusion, no matter what you are doing in a reference job, it will never appear in the job is ultimately run.
> Here is the piece of code
> 
> def launch_sphinx_vasp (old_project):
> old_ref=old_project.load("strain_1_0")
> elem=old_ref.initial_structure.elements[0].symbol
> if isinstance(old_ref.input.KpointFolding,Group):
> kpoints= [k for k in old_ref.input.KpointFolding.values()]
> else:
> kpoints=old_ref.input.KpointFolding
> job_name=elem + '_k' + 'x'.join([str(k) for k in kpoints])
> sphinx_vasp_ref = new_project.create_job(new_project.job_type.Sphinx, job_name +'_sphinx_vasp')
> sphinx_vasp_ref.structure = old_ref.initial_structure
> sphinx_abinit_ref.calc_static()
> #print(sphinx_vasp_ref.input.sphinx.to_sphinx())
> sphinx_vasp_ref.input['KpointCoords'] = [0.0, 0.0, 0.0]
> sphinx_vasp_ref.set_kpoints(kpoints)
> sphinx_vasp_ref.set_encut(old_ref.encut)
> sphinx_vasp_ref.input['EmptyStates'] = old_ref.input['EmptyStates']
> sphinx_vasp_ref.input['KJxc'] = True
> sphinx_vasp_ref.server.queue = 'cm'
> sphinx_vasp_ref.server.cores = old_ref.server.cores
> sphinx_vasp_ref.set_occupancy_smearing(width=0.1)
> sphinx_vasp_ref.input.sphinx.pawPot.species={"checkOverlap":0}
> 
> ```
> ########
> murn_sphinx_vasp = new_project.create_job(new_project.job_type.Murnaghan, job_name+str("_sphinx_vasp_murn"))
> murn_sphinx_vasp.input['vol_range'] = 0.06
> murn_sphinx_vasp.input['num_points'] = 7
> murn_sphinx_vasp.ref_job = sphinx_vasp_ref
> murn_sphinx_vasp.run()
> ```
> 
> One speculation is that the sphinx job is copied, loses the sphinx specific information, or at the way; jobs are set up in Murnaghan job are not compatible with coping sphinx job.
> @ashtonmv @freyso

I'm struggling...

Can you say which lines in your code are not working?

**Comment by mostafajoulaian** [comment]
sphinx_vasp_ref.input.sphinx.pawPot.species={"checkOverlap":0}
This line checkOverlap

**Comment by freyso** [comment]
This line is working as expected, I suppose. What is not working is to transfer this setting to the murn_sphinx_vasp job created here:
murn_sphinx_vasp = new_project.create_job(new_project.job_type.Murnaghan, job_name+str("_sphinx_vasp_murn"))
and assigned the reference job here:
murn_sphinx_vasp.ref_job = sphinx_vasp_ref
I further suspect that the actual job copying happens here:
murn_sphinx_vasp.run()
but Michael and me were not able to trace back where this function ultimately ends up in the pyiron core source code, due to multiple overloads.

**Comment by samwaseda** [comment]
What happens if you write
```
murn_sphinx_vasp = sphinx_vasp_ref.create_job(new_project.job_type.Murnaghan, job_name+str("_sphinx_vasp_murn"))
```

instead of
```
murn_sphinx_vasp = new_project.create_job(new_project.job_type.Murnaghan, job_name+str("_sphinx_vasp_murn"))
murn_sphinx_vasp.ref_job = sphinx_vasp_ref
```
?

**Comment by samwaseda** [comment]
> sphinx_vasp_ref.input.sphinx.pawPot.species={"checkOverlap":0}
> This line checkOverlap

I just checked it myself and it appears to me that it's working, as @freyso also pointed out.

**Comment by mostafajoulaian** [comment]
> What happens if you write
> 
> ```
> murn_sphinx_vasp = sphinx_vasp_ref.create_job(new_project.job_type.Murnaghan, job_name+str("_sphinx_vasp_murn"))
> ```
> 
> instead of
> 
> ```
> murn_sphinx_vasp = new_project.create_job(new_project.job_type.Murnaghan, job_name+str("_sphinx_vasp_murn"))
> murn_sphinx_vasp.ref_job = sphinx_vasp_ref
> ```
> 
> ?

I used the line that you mention but still my job getting aborted and when I look at the file  I see "
<img width="970" alt="Screenshot 2020-12-15 at 10 37 07" src="https://user-images.githubusercontent.com/62750893/102199300-c172c080-3ec3-11eb-8962-98de966740f2.png">
<img width="1187" alt="Screenshot 2020-12-15 at 10 53 15" src="https://user-images.githubusercontent.com/62750893/102199322-c6d00b00-3ec3-11eb-974b-c1b5491ec791.png">

**Comment by freyso** [comment]
Hi Mostafa,
these are just the general error messages of pyiron, which don't show the problem. Can you post the content of the pawPot{} group in the input.sx file that is being generated? I guess the "checkOverlap=0;" item is still missing...

**Comment by mostafajoulaian** [comment]
file which is created does not include "input.sx" which I don't know why but for the time being I look the input file through pyiron
<img width="1078" alt="Screenshot 2020-12-15 at 11 04 50" src="https://user-images.githubusercontent.com/62750893/102200781-912c2180-3ec5-11eb-8e9f-fe0df33582b4.png">

**Comment by samwaseda** [comment]
@mostafajoulaian post `print(a.input.to_sphinx())`

**Comment by mostafajoulaian** [comment]
CheckOverlap = 1;
Ediff = 0.0001;
EmptyStates = 14;
EnCut = 600;
Estep = 100;
KJxc = 1;
KpointCoords = 0.0;
KpointCoords = 0.0;
KpointCoords = 0.0;
KpointFolding = 21;
KpointFolding = 21;
KpointFolding = 21;
SaveMemory = 1;
Sigma = 0.1;
THREADS = 1;
VaspPot = 1;
WriteWaves = 1;
Xcorr = PBE;
rhoMixing = 1.0;
sphinx {
	PAWHamiltonian {
		ekt = 0.1;
		nEmptyStates = 14;
		spinPolarized = 0;
		xc = PBE;
	}
	basis {
		eCut = 44.09918661078599;
		folding = [21, 21, 21];
		kPoint {
			coords = [0.0, 0.0, 0.0];
			relative = 1;
			weight = 1;
		}
		saveMemory = 1;
	}
	initialGuess {
		noWavesStorage = 0;
		rho {
			atomicOrbitals = 1;
		}
		waves {
			lcao {}
			pawBasis = 1;
		}
	}
	main {
		evalForces {
			file = "relaxHist.sx";
		}
		scfDiag {
			blockCCG {}
			dEnergy = 3.674932217565499e-06;
			maxSteps = 100;
			rhoMixing = 1.0;
			spinMixing = 1.0;
		}
	}
	pawPot {
		kjxc = 1;
		species {
			element = "He";
			name = "He";
			potType = "VASP";
			potential = "He_POTCAR";
		}
	}
	structure {
		cell = [[5.3396307128551435, 0.0, 0.0], [2.6698153564275717, 4.624255849689206, 0.0], [5.339198541049896e-16, 3.082587706986659e-16, 8.719572928653767]];
		species {
			atom {
				coords = [0.0, 0.0, 0.0];
			}
			atom {
				coords = [2.6697886638387645, 1.5414032055309344, 4.359786464326883];
			}
			element = "He";
		}
	}
}
spinMixing = 1.0;

**Comment by samwaseda** [comment]
well this time it would have been easier to see a picture instead of text...

**Comment by mostafajoulaian** [comment]
![Uploading Screenshot 2020-12-15 at 11.10.54.png…]()

**Comment by mostafajoulaian** [comment]
<img width="581" alt="Screenshot 2020-12-15 at 11 10 54" src="https://user-images.githubusercontent.com/62750893/102201406-5971a980-3ec6-11eb-83a6-64093d47ad9b.png">

**Comment by freyso** [comment]
Yes, exactly. The pawPot{} group is not set up as intended and specified in the ref_job.
@mostafajoulaian : please avoid pasting screen shots when copying text is sufficient.

**Comment by freyso** [comment]
I think I found the place where and how the reference job is copied to the single instances:
pyiron_base/pyiron_base/master/parallel.py , line 817:
`        job = self.ref_job.copy()`
If I correctly identified this, the error would be in the copy() function for sphinx jobs.

**Comment by freyso** [comment]
I tried to find out where this copy() function goes, but failed.

However, I discovered that the SPHInX job class contains an interface to checkOverlap via
`job.input.CheckOverlap = False`
and apparently also supports JTH GGA type potentials (including copying) via
`job.input.VaspPot = False`
maybe these settings are being copied? Mostafa, can you please try?

**Comment by samwaseda** [comment]
I still haven't got the problem properly. Can you maybe briefly explain what was expected and what you got instead? By the way, this was included in the template, so it would be nice to not erase it when you open a new PR, as pointed out by @jan-janssen recently...

**Comment by freyso** [comment]
> Can you maybe briefly explain what was expected and what you got instead?
We make the following setting in the reference job:
sphinx_vasp_ref.input.sphinx.pawPot.species={"checkOverlap":0}

This should set the "checkOverlap" entry in the pawPot group to 0 (=false), and this would work if we ran the reference job. 
WHAT I EXPECTED was that this setting is copied from the reference job to the 7 murn jobs at different strains that are created by the Murnaghan job class.
WHAT I GOT INSTEAD: However, when copying the reference job to the 7 murn jobs at different strains, all these direct settings are lost. This happens for everything we set in the new "direct access to sx input file via the new "Group" class, so it is a general problem in the future, because we will need to play with pyiron-unsupported SPHInX features. For the present case, we might try setting this via pyiron's "sphinx wrapper class" attributes that cover part of the possible input file settings, and hope that this works as expected (this is what I asked Mostafa to try out).

> By the way, this was included in the template, so it would be nice to not erase it when you open a new PR, as pointed out by @jan-janssen recently...
I don't really get this.

**Comment by ashtonmv** [comment]
I have a minimum (standalone) working example below that I think explains the problem for @samwaseda as well as solves it for @mostafajoulaian:

@samwaseda the problem was that the input settings for `checkOverlap`, `potType`, etc. were being set correctly when we created the reference job, but then did not make it down to the murnaghan children jobs' inputs. So the 2nd print statement I have below was returning not the same pawPot group as the first print statement. With the code I have below, however, it is. I hope that makes sense.

```python
from pyiron import Project


new_project = Project("test")

element = "Fe"

sphinx_vasp_ref = new_project.create_job(new_project.job_type.Sphinx, "murn_test_ref")
sphinx_vasp_ref.structure = new_project.create_structure(element, 'bcc', 2.55)
sphinx_vasp_ref.load_default_groups()

sphinx_vasp_ref.set_kpoints(mesh=[13, 13, 13], center_shift=[0, 0, 0])
sphinx_vasp_ref.set_encut(400)
sphinx_vasp_ref.set_empty_states(10)

# Setting this via set_occupancy_smearing(width=0.1) is apparently ignored...?
sphinx_vasp_ref.input.sphinx.PAWHamiltonian.ekt = 0.1

sphinx_vasp_ref.input.sphinx.pawPot.kjxc = True
sphinx_vasp_ref.input.sphinx.pawPot.species[0].potType = "VASP"
sphinx_vasp_ref.input.sphinx.pawPot.species[0].potential = f"{element}_POTCAR"
sphinx_vasp_ref.input.sphinx.pawPot.species.checkOverlap = 0

sphinx_vasp_ref.server.queue = "s_cmfe"
sphinx_vasp_ref.server.cores = 4

print(sphinx_vasp_ref.input.sphinx.to_sphinx())
```
```
.
.
.
pawPot {
	species {
		0 {
			name = "Fe";
			potType = VASP;
			element = "Fe";
			potential = Fe_POTCAR;
		}
		checkOverlap = 0;  # looks good
	}
	kjxc;
}
.
.
.
```
```python
murn_sphinx_vasp = new_project.create_job(new_project.job_type.Murnaghan, "murn_test")
murn_sphinx_vasp.input['vol_range'] = 0.06
murn_sphinx_vasp.input['num_points'] = 7
murn_sphinx_vasp.ref_job = sphinx_vasp_ref
murn_sphinx_vasp.run(run_again=True)

print(new_project.load(murn_sphinx_vasp.child_ids[0]).input.sphinx.to_sphinx())
```

```
.
.
.
pawPot {
	kjxc = 1;
	species {
		0 {
			element = "Fe";
			name = "Fe";
			potType = VASP;
			potential = Fe_POTCAR;
		}
		checkOverlap = 0;  # wasn't here before, that was the problem
	}
}
.
.
.
```

So long story short I don't think this is a bug after all, but rather some mixup in what we had originally written. I'm not sure exactly where, but I'm suspicious of lines like this:

```python
sphinx_vasp_ref.structure = old_ref.initial_structure
sphinx_abinit_ref.calc_static()
```

**Comment by samwaseda** [comment]
> This should set the "checkOverlap" entry in the pawPot group to 0 (=false), and this would work if we ran the reference job.
> WHAT I EXPECTED was that this setting is copied from the reference job to the 7 murn jobs at different strains that are created by the Murnaghan job class.
> WHAT I GOT INSTEAD: However, when copying the reference job to the 7 murn jobs at different strains, all these direct settings are lost. This happens for everything we set in the new "direct access to sx input file via the new "Group" class, so it is a general problem in the future, because we will need to play with pyiron-unsupported SPHInX features. For the present case, we might try setting this via pyiron's "sphinx wrapper class" attributes that cover part of the possible input file settings, and hope that this works as expected (this is what I asked Mostafa to try out).

Oooooooooooh ok now I got it. Let me check it then. But is @ashtonmv saying it's not a problem in the end?

> By the way, this was included in the template, so it would be nice to not erase it when you open a new PR, as pointed out by @jan-janssen recently...
> I don't really get this.

This one is related to [this PR](https://github.com/pyiron/pyiron/issues/1254).

**Comment by freyso** [comment]
@ashtonmv :  I assume the magic is here:
`sphinx_vasp_ref.load_default_groups()`
because this creates the internal representation of the input file such that it can be reliably modified and copied. Is there any tutorial/example/reference where we could document this?

@samwaseda , I'll apologize for the confusion and effort you may have put into this. I'll go through this report with @mostafajoulaian to explain to him how efficient bug reporting works.

**Comment by ashtonmv** [comment]
> @ashtonmv : I assume the magic is here:
sphinx_vasp_ref.load_default_groups()
because this creates the internal representation of the input file such that it can be reliably modified and copied. Is there any tutorial/example/reference where we could document this?

I think you're right, but `calc_static()` calls `load_default_groups()` internally, which is why I'm suspicious that the original problem was that we actually called calc_static on `sphinx_abinit_ref` while we were working on `sphinx_vasp_ref`...

**Comment by freyso** [comment]
> I think you're right, but calc_static() calls load_default_groups() internally, which is why I'm suspicious that the original problem was that we actually called calc_static on sphinx_abinit_ref while we were working on sphinx_vasp_ref...

alas! stupid copy&paste error. As classic as it could get...

Nonetheless, we probably should turn your working example into an "non-generic SPHInX settings" tutorial, where we explain all this. What do you think?

**Comment by samwaseda** [comment]
So if it's just about `checkOverlap`, I restored the function `job.set_check_overlap` [here](https://github.com/pyiron/pyiron/pull/1263), which will overwrite everything (and will not be overwritten by other functions), and since this one is part of the generic input (which is not even a little bit generic though), it should always be copied.

**Comment by ashtonmv** [comment]
@freyso Yes, I completely agree. I also discovered a few other surprises along the way that I think might be worth documenting/ patching somewhere.

And thanks for humoring us and for your help, @samwaseda 😄  I think we also had the same issue with flags/variables other than checkOverlap, but please don't worry about those. My hope with the Group access was that loose-cannon users could set some of these more "exotic" flags without the pyiron devs ever having to include them in the hard-coded input heirarchy.

I guess we can close this?

**Comment by freyso** [comment]
I close this, as our original claim that copying does not work properly has turned out to be a combination of stupid mistakes and misunderstandings of how to impose SPHInX settings on Murnaghan jobs.
The fix for set_check_overlap () is still a good outcome (Thanks, @samwaseda), and so will be the SPHInX settings tutorial (thanks, @ashtonmv)!
